### PR TITLE
chore: add versioning and allowdowngrade flag for pipeline template

### PR DESCRIPTION
**Pull request by iamhopaul123**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/cli/pipeline_deploy.go [hunk 1]
```diff
@@ -211,8 +224,41 @@ func (o *deployPipelineOpts) Ask() error {
...
+	pipelineVersion, err := vg.Version()
+	if err != nil {
+		// If the stack doesn't exist, exit gracefully.
+		var errStackNotExist *cloudformation.ErrStackNotFound
+		if errors.As(err, &errStackNotExist) {
+			return nil
+		}
+		return fmt.Errorf("get template version of pipeline %s: %w", name, err)
+	}
+	if diff := semver.Compare(pipelineVersion, templateVersion); diff > 0 {
```
Comment (dannyrandall) [inline internal/pkg/cli/pipeline_deploy.go]:
```suggestion
	if semver.Compare(pipelineVersion, templateVersion) > 0 {
```

File: internal/pkg/describe/workload.go [hunk 2]
```diff
@@ -63,7 +63,11 @@ func NewWorkloadStackDescriber(opt NewWorkloadConfig) (*WorkloadStackDescriber,
 //
 // If the Version field does not exist, then it's a legacy template and it returns an version.LegacyWorkloadTemplate and nil error.
 func (d *WorkloadStackDescriber) Version() (string, error) {
-	raw, err := d.cfn.StackMetadata()
+	return stackVersion(d.cfn, version.LegacyWorkloadTemplate)
+}
+
+func stackVersion(descr stackDescriber, legacyVersion string) (string, error) {
```
Comment (dannyrandall) [inline internal/pkg/describe/workload.go:69]: nice! 👏

**Review by dannyrandall — APPROVED** [review]

File: internal/pkg/cli/errors.go [hunk 3]
```diff
@@ -9,6 +9,29 @@ import (
...
+	version         string
+	templateVersion string
+}
+
+func (e *errCannotDowngradePipelineVersion) init() *errCannotDowngradeVersion {
+	return &errCannotDowngradeVersion{
+		componentName:         e.name,
+		componentType:         "pipeline",
+		currentVersion:        e.version,
+		latestTemplateVersion: e.templateVersion,
```
Comment (Lou1415926) [inline internal/pkg/cli/errors.go]: nit nit nit: feel free to ignore this if you disagree! I felt like ⬇️  naming might be clearer...? `currentVersion` could mean "the current deployed template version" or "the current latest version", and `latestTemplateVersion` could mean the latest version of all time, or the latest version of "this release". Anyway, after not reviewing PRs for a while, I got a little confused with the naming 😂 hence the suggesetion. But again, this is probably just me, so feel free to ignore.
```go
type errCannotDowngradeVersion struct {
   laterVersion string // previously `currentVersion`.
   thisVersion string // previously `latestTemplateVersion`.
}
```

File: internal/pkg/cli/pipeline_deploy.go [hunk 4]
```diff
@@ -211,8 +224,41 @@ func (o *deployPipelineOpts) Ask() error {
 	return o.askWsPipelineName()
 }
 
+func validatePipelineVersion(vg versionGetter, name, templateVersion string) error {
+	pipelineVersion, err := vg.Version()
+	if err != nil {
+		// If the stack doesn't exist, exit gracefully.
```
Comment (Lou1415926) [inline internal/pkg/cli/pipeline_deploy.go]: nit:
```suggestion
```

**Review by huanjani — APPROVED** [review]
LGTM!

**Review by Lou1415926 — APPROVED** [review]
LGTM! Just some nits

**Pull request merged**